airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py

```diff
@@ -0,0 +1,151 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from contextlib import contextmanager
+from io import IOBase
+from typing import Any, Dict, Iterable, List, Optional, Set
+
+import boto3.session
+import smart_open
+from botocore.client import BaseClient, Config
+from airbyte_cdk.sources.file_based.exceptions import ErrorListingFiles, FileBasedSourceError
+from airbyte_cdk.sources.file_based.file_based_stream_reader import AbstractFileBasedStreamReader
+from airbyte_cdk.sources.file_based.remote_file import FileReadMode, RemoteFile
+from source_s3.v4.config import Config
+
+
+class SourceS3StreamReader(AbstractFileBasedStreamReader):
+
+    def __init__(self):
+        super().__init__()
+        self._s3_client = None
+
+    @property
+    def config(self) -> Config:
+        return self._config
+
+    @config.setter
+    def config(self, value: Config):
+        assert isinstance(value, Config)
+        self._config = value
+
+    @property
+    def s3_client(self) -> BaseClient:
+        if self.config is None:
+            # We shouldn't hit this; config should always get set before attempting to
+            # list or read files.
+            raise ValueError("Source config is missing; cannot create the S3 client.")
+        if self._s3_client is None:
+            if self.config.endpoint:
+                client_kv_args = _get_s3_compatible_client_args(self.config)
+                self._s3_client = boto3.client('s3', **client_kv_args)
+            else:
+                self._s3_client = boto3.client(
+                    's3',
+                    aws_access_key_id=self.config.aws_access_key_id,
+                    aws_secret_access_key=self.config.aws_secret_access_key,
+                )
+        return self._s3_client
+
+    def get_matching_files(self, globs: List[str], logger: logging.Logger) -> Iterable[RemoteFile]:
+        """
+        Get all files matching the specified glob patterns.
+        """
+        s3 = self.s3_client
+        prefixes = self.get_prefixes_from_globs(globs)
+        seen = set()
+        total_n_keys = 0
+
+        try:
+            kwargs = {"Bucket": self.config.bucket}
+            if prefixes:
+                for prefix in prefixes:
+                    for remote_file in self._page(s3, globs, kwargs, prefix, seen, logger):
+                        yield remote_file
+            else:
+                total_n_keys += 1
+                for remote_file in self._page(s3, globs, kwargs, None, seen, logger):
+                    yield remote_file
+
+            logger.info(f"Finished listing objects from S3. Found {total_n_keys} objects total ({len(seen)} unique objects).")
```

Comment:
Although `len(seen)` would be correct for both flows, wouldn't `total_n_keys` always be 1 if there are no prefixes. And for prefixes, it would always log 0 since it never changes `total_n_keys`. We count `total_n_keys_for_prefix` within `_page()` but we only print the extra log in that method. 

What count are we trying to display here and would this extra log be duplicative with the ones in `_page()`?